src/client/Microsoft.Identity.Client/OAuth2/Throttling/UiRequiredProvider.cs

```diff
@@ -0,0 +1,118 @@
+        }
+
+        public void ResetCache()
+        {
+            ThrottlingCache.Clear();
+        }
+
+        public void TryThrottle(AuthenticationRequestParameters requestParams, IReadOnlyDictionary<string, string> bodyParams)
+        {
+            if (!ThrottlingCache.IsEmpty() && IsSupported(requestParams))
+            {
+                var logger = requestParams.RequestContext.Logger;
+
+                string fullThumbprint = GetRequestStrictThumbprint(
+                    bodyParams,
+                    requestParams.AuthorityInfo.CanonicalAuthority,
+                    requestParams.RequestContext.ServiceBundle.PlatformProxy.CryptographyManager);
+
+                TryThrowException(fullThumbprint, logger);
+            }
+        }
+
+        private void TryThrowException(string thumbprint, ICoreLogger logger)
+        {
+            if (ThrottlingCache.TryGetOrRemoveExpired(thumbprint, logger, out MsalServiceException ex) &&
+                ex is MsalUiRequiredException uiException)
+            {
+                logger.WarningPii(
+                    $"[Throttling] Exception thrown because of throttling rule UiRequired - thumbprint: {thumbprint}",
+                    $"[Throttling] Exception thrown because of throttling rule UiRequired ");
+
+                // mark the exception for logging purposes                
+                throw new MsalThrottledUiRequiredException(uiException);
+            }
+        }
+
+        private static bool IsSupported(AuthenticationRequestParameters requestParams)
+        {            
+            return !requestParams.IsConfidentialClient &&
+                requestParams.ApiId == TelemetryCore.Internal.Events.ApiEvent.ApiIds.AcquireTokenSilent;
+        }
+
+        /// <summary>
+        /// The strict thumbprint is based on: 
+        /// ClientId
+        /// Authority (env + tenant)
+        /// Scopes
+        /// hash(RT) or UPN for IWA (not supported)
+        /// </summary>
+        private static string GetRequestStrictThumbprint(
+            IReadOnlyDictionary<string, string> bodyParams,
+            string authority,
+            ICryptographyManager crypto)
+        {
+            StringBuilder sb = new StringBuilder();
+            if (bodyParams.TryGetValue(OAuth2Parameter.ClientId, out string clientId))
+            {
+                sb.Append((clientId ?? "") + ThrottleCommon.KeyDelimiter);
+            }
+            sb.Append(authority + ThrottleCommon.KeyDelimiter);
+            if (bodyParams.TryGetValue(OAuth2Parameter.Scope, out string scopes))
+            {
+                sb.Append((scopes ?? "") + ThrottleCommon.KeyDelimiter);
+            }
+
+            if (bodyParams.TryGetValue(OAuth2Parameter.RefreshToken, out string rt) && 
+                !string.IsNullOrEmpty(rt))
+            {
+                sb.Append(crypto.CreateSha256Hash(rt) + ThrottleCommon.KeyDelimiter);
+            }
+
+            return sb.ToString();
+        }
+    }
```

Comment:
Does B2C support throttling?

Reply:
Client-side throttling is based on the following signals from the token providres: 

- HTTP 429 and RetryAfter header
- HTTP 5xx
- HTTP 400 and error_code of "invalid_grant" / "interaction_required" (which MSAL translates as MsalUiRequiredException). 

Now, I don't know if B2C sends RetryAfter headers. I think it does send "invalid_grant" errors though. 

Definitely something to follow up with B2C folks, thanks for pointing this out @jennyf19